src/lib/provable/crypto/encryption.ts

```diff
@@ -68,3 +73,108 @@ function decrypt(
 
   return message;
 }
+
+/**
+ * Decrypts a {@link CipherText} using a {@link PrivateKey}.
+ */
+function decryptV2(
+  {
+    publicKey,
+    cipherText,
+    messageLength,
+  }: CipherText & { messageLength: number },
+  privateKey: PrivateKey
+) {
+  // key exchange
+  const sharedSecret = publicKey.scale(privateKey.s);
+  const sponge = new Poseidon.Sponge();
+  sponge.absorb(sharedSecret.x);
+  const authenticationTag = cipherText.pop();
+
+  // decryption
+  const message = [];
+  for (let i = 0; i < cipherText.length; i++) {
+    // absorb frame tag
+    if (i === cipherText.length - 1) sponge.absorb(Field(1));
+    else sponge.absorb(Field(0));
+
+    const keyStream = sponge.squeeze();
+    const messageChunk = cipherText[i].sub(keyStream);
+
+    // convert to bytes
+    const byteMessage = wordToBytes(messageChunk, 32);
+
+    // push the message to our final message array
+    message.push(byteMessage);
+
+    if (i % 2 === 1) sponge.absorb(cipherText[i - 1]);
+    if (i % 2 === 1 || i === cipherText.length - 1)
+      sponge.absorb(cipherText[i]);
+  }
+
+  // authentication tag
+  sponge.squeeze().assertEquals(authenticationTag!);
+
+  // calculate padding
+  const multipleOf = 31;
+  const n = Math.ceil(messageLength / multipleOf) * multipleOf;
+
+  // return the message as a flat array of bytes, slice the padding off of the final message
+  return Bytes.from(message.flat().slice(0, messageLength - n));
+}
+
+/**
+ * Public Key Encryption, encrypts Bytes using a {@link PublicKey}.
+ */
+function encryptV2(
+  message: Bytes,
+  otherPublicKey: PublicKey
+): CipherText & {
+  messageLength: number;
+} {
+  const bytes = message.bytes;
+  const messageLength = bytes.length;
+
+  // pad message to a multiple of 31 so they still fit into one field element
+  const multipleOf = 31;
+  const n = Math.ceil(messageLength / multipleOf) * multipleOf;
+
+  // create the padding
+  const padding = Array.from({ length: n - messageLength }, () =>
+    UInt8.from(0)
+  );
+  message.bytes = bytes.concat(padding);
+
+  // convert message into chunks of 31 bytes
+  const chunks = message.chunk(31);
+
+  // key exchange
+  const privateKey = Provable.witness(Scalar, () => Scalar.random());
+  const publicKey = Group.generator.scale(privateKey);
+  const sharedSecret = otherPublicKey.toGroup().scale(privateKey);
+
+  const sponge = new Poseidon.Sponge();
+  sponge.absorb(sharedSecret.x);
+
+  // encryption
+  const cipherText = [];
+  for (let [n, chunk] of chunks.entries()) {
+    // absorb frame bit
+    if (n === chunks.length - 1) sponge.absorb(Field(1));
+    else sponge.absorb(Field(0));
```

Comment:
ok so now you're simply absorbing the frame bit instead of concatenating it with the message!

in that case,
- the optimization done below to always absorb two ciphertext chunks at a time becomes obsolete, and slower than absorbing them one at a time, because including the frame bit we're doing 2 at a time anyway. so remove that optimization i.e. just `sponge.absorb(encryptedChunk)` below
- since we're using an entire separate field element for the frame bit, we could now again have a version of `encryptV2()` which operates on messages of full field elements. and which doesn't have the issue that the message length is ambiguous.
  - consider this as an optional refactor. in that case, the concatenation of bytes to fields could be a separate step done before calling encrypt to give an alternative API with `Bytes` input, which needs to add the message length to the ciphertext output.

Reply:
Yea, absorbing it directly makes a little more sense now that I got more familiar with the sponge concept, its also a little easier and we can indeed go back to field elements - I think I will do that